File: 7. breadth-first-search/list-queue.py
# python3

from collections import deque

graph = {}
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

def search(name):
  search_queue = deque() # create a queue
  search_queue += graph[name] # add all your neighbours (alice, bob and claire) to the queue.
  searched = []

  while search_queue: # While the queue isn't empty
    person = search_queue.popleft() # dequeue the first person from queue.
    if not person in searched:
      if person_is_seller(person):
        print(person + " is a mango seller!")
        return True
      else:
        search_queue += graph[person] # Add all of this person's friends to the search queue.
        searched.append(person)
  return False

# This function checks whether the person's name ends with the letter m. If it does, they're a mango seller.
def person_is_seller(name):
  return name[-1] == 'm'

search("you")

File: 12-breadth-first-search/list-queue.rb
# encoding: UTF-8

require 'thread'

@graph = [["alice", "bob", "claire"], ["anuj", "peggy"], ["peggy"], ["thom", "jonny"]]
@search_queue = Queue.new

def enqueue(arr)
  added = []

  arr.each do |item|
    item.each do |element|
      unless added.include?(element)
        @search_queue.push(element)
        added << element
        puts "#{element} queued!"
      end
    end
  end
end

def dequeue(queue)
  until queue.empty?
    item = queue.shift
    puts "#{item} dequeued!"
    sleep 0.5
  end
end

def search(name, queue)
  searched = []

  unless queue.empty? || searched.include?(name)
    if person_is_seller(name)
      puts "#{name} is a mango seller!"
      return true
    else
      searched << name
    end
  end
end

# this function checks whether the person's name ends with the letter m. If it does, they're a mango seller.
def person_is_seller(name)
  name[-1] == 't'
end

enqueue(@graph)
dequeue(@search_queue)
enqueue([["serhat"]])
search("serhat", @search_queue)